server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/processors/FileOperationProcess.java

```diff
@@ -90,81 +93,85 @@ private void checkFileOpenedBeforeOperation(
+      ProgramNode programNode,
       String messageTemplate) {
-    node.getNearestParentByType(NodeType.PROGRAM)
-        .ifPresent(
-            pnode -> {
-              boolean isFileOpened =
-                  pnode
-                      .getDepthFirstStream()
-                      .filter(Node.hasType(OPEN_STATEMENT))
-                      .map(OpenStatementNode.class::cast)
-                      .anyMatch(
-                          n ->
-                              n.getFilename().getName().equalsIgnoreCase(filename)
-                                  && expectedFileKind.contains(n.getFileOperationKind()));
-              if (!isFileOpened && Objects.nonNull(errorLocality)) {
-                ctx.getErrors()
-                    .add(
-                        SyntaxError.syntaxError()
-                            .errorSource(ErrorSource.PARSING)
-                            .severity(ErrorSeverity.WARNING)
-                            .location(errorLocality.toOriginalLocation())
-                            .messageTemplate(MessageTemplate.of(messageTemplate))
-                            .build());
-              }
-            });
+
+    if (errorLocality == null) {
+      return Optional.empty();
+    }
+
+    if (!isFileOpen(programNode, filename, expectedFileKind)) {
+      return Optional.of(
+          SyntaxError.syntaxError()
+              .errorSource(ErrorSource.PARSING)
+              .severity(ErrorSeverity.WARNING)
+              .location(errorLocality.toOriginalLocation())
+              .messageTemplate(MessageTemplate.of(messageTemplate))
+              .build());
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isFileOpen(ProgramNode programNode, String filename, List<FileOperationKind> expectedFileKind) {
+    return null != programNode.findFirstNodeInSubtree(n -> {
+        if (n.getNodeType() != OPEN_STATEMENT) {
+          return false;
+        }
+        OpenStatementNode osn = (OpenStatementNode) n;
+        if (!expectedFileKind.contains(osn.getFileOperationKind())) {
+          return false;
+        }
+        return osn.getFilename().getName().equalsIgnoreCase(filename);
+    });
   }
 
   private void checkFileOpenedForWrite(
       FileOperationStatementNode node,
       List<FileOperationKind> expectedFileKind,
       ProcessingContext ctx,
       String messageTemplate) {
-    List<QualifiedReferenceNode> listOfWriteVariables =
-        node.getChildren().stream()
-            .filter(n -> node.getFilename().getLocality().equals(n.getLocality()))
-            .filter(Node.hasType(NodeType.QUALIFIED_REFERENCE_NODE))
-            .map(QualifiedReferenceNode.class::cast)
-            .collect(Collectors.toList());
-    if (listOfWriteVariables.size() != 1) {
-      return;
-    }
-
-    QualifiedReferenceNode writeVariable = listOfWriteVariables.get(0);
-
-    writeVariable
+      List<QualifiedReferenceNode> listOfWriteVariables = new ArrayList<>();
```

Comment:
We don't need a list here as well. Just return if a second instance is found or none exists.